lints/lint_san_dns_name_onion_not_ev_cert.go

```diff
@@ -0,0 +1,78 @@
+/*
+ * ZLint Copyright 2019 Regents of the University of Michigan
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy
+ * of the License at http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package lints
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/zmap/zcrypto/x509"
+	"github.com/zmap/zlint/util"
+)
+
+const (
+	// onionTLD is a const for the TLD for Tor Hidden Services.
+	onionTLD = ".onion"
+)
+
+type onionNotEV struct{}
+
+// Initialize for an onionNotEV linter is a NOP.
+func (l *onionNotEV) Initialize() error {
+	return nil
+}
+
+// CheckApplies returns true if the certificate is a subscriber certificate that
+// contains a subject name ending in `.onion`.
+func (l *onionNotEV) CheckApplies(c *x509.Certificate) bool {
+	return util.IsSubscriberCert(c) && util.CertificateSubjInTLD(c, onionTLD)
+}
+
+// Execute checks that if the certificate contains any subject names ending in
+// the  .onion TLD that it is also an EV certificate. If there are .onion names
+// and the certificate is not an EV certificate an Error LintResult is returned.
+func (l *onionNotEV) Execute(c *x509.Certificate) *LintResult {
+	isEV := util.IsEV(c.PolicyIdentifiers)
+	names := append(c.DNSNames, c.Subject.CommonName)
+	for _, name := range names {
+		name = strings.ToLower(name)
+		/*
+		 * Effective May 1, 2015, each CA SHALL revoke all unexpired Certificates with an
+		 * Internal Name using onion as the right-most label in an entry in the
+		 * subjectAltName Extension or commonName field unless such Certificate was
+		 * issued in accordance with Appendix F of the EV Guidelines.
+		 */
+		if strings.HasSuffix(name, onionTLD) && !isEV {
+			return &LintResult{
+				Status: Error,
+				Details: fmt.Sprintf(
+					"certificate contains %s subject domain (%q) but is not an EV certificate",
+					onionTLD, name),
+			}
+		}
+	}
+	return &LintResult{Status: Pass}
+}
+
+func init() {
+	RegisterLint(&Lint{
+		Name:          "san_dns_name_onion_not_ev_cert",
+		Description:   "certificates with a .onion subject name must be issued in accordance with EV Guidelines",
+		Citation:      "CABF Ballot 144",
```

Comment:
I left this lint citing Ballot 144 despite [feedback to do otherwise](https://github.com/zmap/zlint/pull/263#discussion_r261844580).

[Ballot 144](https://cabforum.org/2015/02/18/ballot-144-validation-rules-dot-onion-names/) added the text that I think most directly maps to this linter. It updated section 9.2.1 ("Subject Alternative Name Extension") of the BRs (v1.2.3 at the time of Ballot 144) to add:
> Effective May 1, 2015, each CA SHALL revoke all unexpired Certificates with an Internal Name using onion as the right-most label in an entry in the subjectAltName Extension or commonName field unless such Certificate was issued in accordance with Appendix F of the EV Guidelines.

I'm not sure when that text was lost, because in the current BRs ([v1.6.3](https://cabforum.org/wp-content/uploads/CA-Browser-Forum-BR-1.6.3.pdf) at the time of writing) I can't find _any_ mention of `.onion` except in section 1.2.1. which is just a table of revisions. Looking at section 7.1.4.2.1 ("Subject Alternative Name Extension") of BRs v1.6.3 - the section I think maps closest to v1.2.3's equivalent section -  there is no longer any mention of `.onion`.

Perhaps there's an argument to be made that the correct citation is BRs section 7.1.4.2.1 because that section forbids internal names in a SAN. The v1.6.3 BRs define an "internal name" in Section 1.6.1 "Definitions" as:
> A string of characters (not an IP address) in a Common Name or Subject Alternative Name field of a Certificate that cannot be verified as globally unique within the public DNS at the time of certificate issuance because it does not end with a Top Level Domain registered in IANA’s Root Zone Database.

The `.onion` top level domain isn't delegated from the root zonein the public DNS so it seems as though by the BRs alone a `.onion` name would be forbidden in a SAN. The part that seems missing here is the connection from being forbidden as an internal name in a non-EV cert to being allowed in the EV guidelines. Section 11.7.1 ("Verification Requirements") of the EV guidelines explicitly allows `.onion` provided the rules in Appendix F are followed.

Phew! That's a lot of words. The tl;dr is that I think citing the ballot makes the most sense but if someone has a better picture of the history and prefers an alternative citation I'd be happy to oblige.